### PR TITLE
Add PubSub dead letter topic support

### PR DESCRIPTION
**Pull request by Taction**
Signed-off-by: zhangchao <zchao9100@gmail.com>

# Description

Add PubSub's DeadLetter Topic support.

## Issue reference

Please reference the issue this PR will close: #2217

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: pkg/runtime/runtime.go [hunk 1]
```diff
@@ -94,6 +94,7 @@ const (
 	bindingsConcurrencyParallel   = "parallel"
 	bindingsConcurrencySequential = "sequential"
 	pubsubName                    = "pubsubName"
+	deadLetterKeyFormat           = "%s||%s" // componentName||topicName
```
Comment (yaron2) [inline pkg/runtime/runtime.go:97]: You might have two different components with the same name, deployed to different namespaces:

mypubsub deployed to ns1, and mypubsub deployed to ns2.
Does this need to be changed to include a namespace?

Comment (Taction) [reply pkg/runtime/runtime.go:97]: Yes, I'll fix it.

Comment (Taction) [reply pkg/runtime/runtime.go:97]: After a view of the code, components in the different namespaces will be ignored. https://github.com/dapr/dapr/blob/master/pkg/runtime/runtime.go#L1754

**Comment by yaron2** [comment]
@Taction please resolve conflicts. This is a great PR and want to get this into 1.8!

**Comment by daixiang0** [comment]
@Taction is it ready for review as I see WIP in the title?

**Comment by Taction** [comment]
@daixiang0 It is ready for review. I am adding unit and e2e tests for it.

File: pkg/runtime/pubsub/subscription.go [hunk 2]
```diff
@@ -1,11 +1,12 @@
...
 
 type Subscription struct {
-	PubsubName string            `json:"pubsubname"`
-	Topic      string            `json:"topic"`
-	Metadata   map[string]string `json:"metadata"`
-	Rules      []*Rule           `json:"rules,omitempty"`
-	Scopes     []string          `json:"scopes"`
+	PubsubName      string            `json:"pubsubname"`
+	Topic           string            `json:"topic"`
+	DeadLetterTopic string            `json:"dead_letter_topic"`
```
Comment (yaron2) [inline pkg/runtime/pubsub/subscription.go:6]: Should be `deadLetterTopic` to be consistent with the subscription CRD definition.

File: pkg/runtime/pubsub/subscriptions.go [hunk 3]
```diff
@@ -41,11 +41,12 @@ const (
...
 type (
 	SubscriptionJSON struct {
-		PubsubName string            `json:"pubsubname"`
-		Topic      string            `json:"topic"`
-		Metadata   map[string]string `json:"metadata,omitempty"`
-		Route      string            `json:"route"`  // Single route from v1alpha1
-		Routes     RoutesJSON        `json:"routes"` // Multiple routes from v2alpha1
+		PubsubName      string            `json:"pubsubname"`
+		Topic           string            `json:"topic"`
+		DeadLetterTopic string            `json:"dead_letter_topic"`
```
Comment (yaron2) [inline pkg/runtime/pubsub/subscriptions.go:46]: Should be `deadLetterTopic` to be consistent with the subscription CRD definition.

**Comment by yaron2** [comment]
@Taction I reviewed, overall this looks good! I put two small comments.

There's one thing we need to ensure before accepting this: that resiliency policies can be used to determine max retries before going to DLQ.

File: pkg/runtime/pubsub/subscription.go (hunk 2 above)
Comment (Taction) [reply pkg/runtime/pubsub/subscription.go]: Done

File: pkg/runtime/pubsub/subscriptions.go (hunk 3 above)
Comment (Taction) [reply pkg/runtime/pubsub/subscriptions.go]: Done

**Comment by Taction** [comment]
@yaron2 Thanks for your review. Fixed.
Resiliency policies can be used to determine max retries before going to DLQ, and I have added a unit test case to ensure that.

**Comment by yaron2** [comment]
@Taction this PR looks good! I'm ready to merge as soon as an e2e test is added.

**Comment by Taction** [comment]
Thanks for your words, I have added it to subscribing app's info. The e2e is in progress, if there is still something wrong I'll look into them tomorrow.

**Comment by Taction** [comment]
@yaron2 e2e test passed.

**Review by yaron2 — CHANGES_REQUESTED** [review]

File: pkg/runtime/runtime.go [hunk 4]
```diff
@@ -588,39 +611,44 @@ func (a *DaprRuntime) beginPubSub(name string, ps pubsub.PubSub) error {
 				data, err = a.json.Marshal(cloudEvent)
 				if err != nil {
 					log.Errorf("error serializing cloud event in pubsub %s and topic %s: %s", name, msg.Topic, err)
-					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Retry)), msg.Topic, 0)
-					return err
+					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
```
Comment (yaron2) [inline pkg/runtime/runtime.go:614]: See comment here: https://github.com/dapr/dapr/pull/4399/files#r861881846.

File: pkg/runtime/runtime.go [hunk 5]
```diff
@@ -588,39 +611,44 @@ func (a *DaprRuntime) beginPubSub(name string, ps pubsub.PubSub) error {
...
+					a.sendToDeadLetterIfConfigured(name, msg)
+					return nil
 				}
 			} else {
 				err = a.json.Unmarshal(msg.Data, &cloudEvent)
 				if err != nil {
 					log.Errorf("error deserializing cloud event in pubsub %s and topic %s: %s", name, msg.Topic, err)
-					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Retry)), msg.Topic, 0)
-					return err
+					diag.DefaultComponentMonitoring.PubsubIngressEvent(ctx, pubsubName, strings.ToLower(string(pubsub.Drop)), msg.Topic, 0)
```
Comment (yaron2) [inline pkg/runtime/runtime.go:622]: See comment here: https://github.com/dapr/dapr/pull/4399/files#r861881846.

**Review by yaron2 — CHANGES_REQUESTED** [review]

**Comment by yaron2** [comment]
@Taction I found one issue across multiple places that needs a small fix in the retry/drop logic. Other than that, LGTM.

**Review by yaron2 — APPROVED** [review]
LGTM.

Great work here @Taction 👍

**Comment by XavierGeerinck** [comment]
❤️ was missing this feature!!!

**Comment by yaron2** [comment]
@Taction there are inconsistent failures with the resiliency + dead letter unit tests. See an example here: https://github.com/dapr/dapr/runs/6284493034?check_suite_focus=true.

Can you please look into why the test is flaky?

**Comment by Taction** [comment]
@yaron2 Sure, I'll look into this.

**Comment by amulyavarote** [comment]
> @yaron2 Sure, I'll look into this.

Thank you so much!

**Pull request merged**